core/kernel/notif.c

```diff
@@ -3,12 +3,217 @@
+static struct notif_driver_head notif_driver_head =
+	SLIST_HEAD_INITIALIZER(&notif_driver_head);
+
+static bitstr_t bit_decl(notif_values, NOTIF_ASYNC_VALUE_MAX + 1);
+static bitstr_t bit_decl(notif_alloc_values, NOTIF_ASYNC_VALUE_MAX + 1);
+static bool notif_started;
+
+TEE_Result notif_alloc_async_value(uint32_t *val)
+{
+	static bool alloc_value_inited;
+	uint32_t old_itr_status = 0;
+	int bit = 0;
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	if (!alloc_value_inited) {
+		bit_set(notif_alloc_values, NOTIF_VALUE_DO_BOTTOM_HALF);
+		alloc_value_inited = true;
+	}
+
+	bit_ffc(notif_alloc_values, (int)NOTIF_ASYNC_VALUE_MAX + 1, &bit);
+	if (bit >= 0) {
+		*val = bit;
+		bit_set(notif_alloc_values, bit);
+	}
+
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+
+	if (bit < 0)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	return TEE_SUCCESS;
+}
+
+void notif_free_async_value(uint32_t val)
+{
+	uint32_t old_itr_status = 0;
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	assert(val < NOTIF_ASYNC_VALUE_MAX);
+	assert(bit_test(notif_alloc_values, val));
+	bit_clear(notif_alloc_values, val);
+
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+}
+
+uint32_t notif_get_value(bool *value_valid, bool *value_pending)
+{
+	uint32_t old_itr_status = 0;
+	uint32_t res = 0;
+	int bit = 0;
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	bit_ffs(notif_values, (int)NOTIF_ASYNC_VALUE_MAX + 1, &bit);
+	*value_valid = (bit >= 0);
+	if (!*value_valid) {
+		*value_pending = false;
+		goto out;
+	}
+
+	res = bit;
+	bit_clear(notif_values, res);
+	bit_ffs(notif_values, (int)NOTIF_ASYNC_VALUE_MAX + 1, &bit);
+	*value_pending = (bit >= 0);
+out:
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+
+	return res;
+}
+
+void notif_send_async(uint32_t value)
+{
+	uint32_t old_itr_status = 0;
+
+	COMPILE_TIME_ASSERT(NOTIF_VALUE_DO_BOTTOM_HALF ==
+			    OPTEE_SMC_ASYNC_NOTIF_VALUE_DO_BOTTOM_HALF);
+	COMPILE_TIME_ASSERT(!CFG_CORE_ASYNC_NOTIF_INTID ||
+			    CFG_CORE_ASYNC_NOTIF_INTID > 31);
+
+	assert(value <= NOTIF_ASYNC_VALUE_MAX);
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	DMSG("0x%"PRIx32, value);
+	bit_set(notif_values, value);
+	itr_raise_pi(CFG_CORE_ASYNC_NOTIF_INTID);
+
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+}
+
+bool notif_async_is_started(void)
+{
+	uint32_t old_itr_status = 0;
+	bool ret;
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+	ret = notif_started;
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+
+	return ret;
+}
+
+void notif_register_driver(struct notif_driver *ndrv)
+{
+	uint32_t old_itr_status = 0;
+	struct notif_driver *nd = NULL;
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	/* Add ndrv at the end of the list */
+	nd = SLIST_FIRST(&notif_driver_head);
+	while (nd && SLIST_NEXT(nd, link))
+		nd = SLIST_NEXT(nd, link);
+	if (nd)
+		SLIST_INSERT_AFTER(nd, ndrv, link);
+	else
+		SLIST_INSERT_HEAD(&notif_driver_head, ndrv, link);
```

Comment:
Why not insert at head? It shouldn't matter as far as I can tell.

Reply:
You're right. I'll fix.